### PR TITLE
Add Tips and Tricks notebook and put examples in the documentation

### PR DESCRIPTION
**Pull request by davebulaval**
(discussion over the #83 lead to this new version of integration of colouring and else)

- added a new notebook for tips and tricks
- The introduction is now lighter (also the README)
- push the colouring, progress_bar, metric name, multi GPU, created callback into a _Tips and tricks_ notebook and web page for the doc.
- standardization of examples notebooks
- typos correction in notebooks
- Added examples section in the doc with the notebooks

### CODE REVIEW

**Comment by davebulaval** [comment]
Todo:

1. Index Tree:
**Exemples**

- Page 1
- Page 2 ...

**API**
-Autres composantes

2. Download notebook before the title.

3. Sortir less commentaires dans notebook et site (pour plus explications)

4. Mettre image de plot the dataset (intro) et pour let's take a look at (tips and trikcs)

5.

File: docs/source/index.rst [hunk 1]
```diff
@@ -140,6 +140,7 @@ Look at notebook files with full working `examples <https://github.com/GRAAL-Res
 - `transfer_learning.ipynb <https://github.com/GRAAL-Research/poutyne/blob/master/examples/transfer_learning.ipynb>`_ - transfer learning on ResNet-18 on the `CUB-200 <http://www.vision.caltech.edu/visipedia/CUB-200-2011.html>`_ dataset.
 - `policy_cifar_example.ipynb <https://github.com/GRAAL-Research/poutyne/blob/master/examples/policy_cifar_example.ipynb>`_ - policies API, FastAI-like learning rate policies
 - `policy_interface.ipynb <https://github.com/GRAAL-Research/poutyne/blob/master/examples/policy_interface.ipynb>`_ - example of policies
+- `tips_and_tricks.ipynb <https://github.com/GRAAL-Research/poutyne/blob/master/examples/tips_and_tricks.ipynb>`_ - example of tips and tricks using Poutyne.
```
Comment (davebulaval) [inline docs/source/index.rst:143]:
```suggestion
- `tips_and_tricks.ipynb <https://github.com/GRAAL-Research/poutyne/blob/master/examples/tips_and_tricks.ipynb>`_ -  tips and tricks using Poutyne.
```

File: examples/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+colorama~=0.4.3
+fasttext~=0.9.2
+requests
+torch
```
Comment (davebulaval) [inline examples/requirements.txt:4]:
```suggestion
colorama~=0.4.3
fasttext~=0.9.2
requests
```

File: docs/source/examples/tips_and_tricks.rst [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+    input_dim = dimension # the output of the LSTM
+    tag_dimension = 8
+
+    fully_connected_network = nn.Linear(input_dim, tag_dimension)
+
+The Dataset
+-----------
+
+Now let's download our dataset; it already split into a train, valid and test set using the following.
```
Comment (freud14) [inline docs/source/examples/tips_and_tricks.rst:100]:
```suggestion
Now let's download our dataset; it's already split into a train, valid and test set using the following.
```

File: docs/source/examples/tips_and_tricks.rst [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+        os.makedirs(saving_dir, exist_ok=True)
+
+        open(os.path.join(saving_dir, f"{data_type}.p"), 'wb').write(r.content)
+
+    download_data('./data/', "train")
+    download_data('./data/', "valid")
+    download_data('./data/', "test")
+
+
+Now let's load in memory the data.
```
Comment (freud14) [inline docs/source/examples/tips_and_tricks.rst:121]:
```suggestion
Now let's load in memory the data.

```
So that the code below renders correctly.

File: docs/source/examples/tips_and_tricks.rst [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+    test_data_vectorize = Vectorizer(test_data, embedding_model)
+
+DataLoader
+^^^^^^^^^^
+
+Now, since all the addresses are not of the same size, it is impossible to batch size them since all elements of a tensor must have the same lengths. But there a trick, padding!
+
+The idea is simple. We will add *empty* tokens at the ends of a sequence up to the longest one in a batch. At the moment of evaluating the loss, that tokens will be skip using a ignore value (`0`). That way, we can pad and pack the sequence to minimize the training time (read `this <https://stackoverflow.com/questions/51030782/why-do-we-pack-the-sequences-in-pytorch>`_ good explanation of why we pad and pack sequence).
+
+Also, due to how the F1 Score to compute is done, we will need a mask to ignore the paddings elements when calculating the metric. The mask value will be set to (`-100`) and will be used only at running time (see the `documentation <https://poutyne.org/metrics.html#poutyne.framework.metrics.FBeta.forward>`_ for more details).
```
Comment (freud14) [inline docs/source/examples/tips_and_tricks.rst:217]: The values in the mask are supposed to be zeros and ones, setting them to -100 will result in some kind of bug... Code to modify below too.

File: docs/source/examples/tips_and_tricks.rst [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+
+For setting those elements, we will use the `collate_fn` of the PyTorch DataLoader, and on running time, that process will be done. We will create a function that will set the padding value (`0`) and the mask value (`-100`).
+
+One time to take into account, since we have packed the sequence, we need the lengths of each sequence for the forward pass to unpack them.
```
Comment (freud14) [inline docs/source/examples/tips_and_tricks.rst:221]:
```suggestion
One thing to take into account, since we have packed the sequence, we need the lengths of each sequence for the forward pass to unpack them.
```

File: docs/source/examples/tips_and_tricks.rst [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+
+            tag_space = self.fully_connected_network(lstm_out)
+            return tag_space.transpose(-1, 1) # we need to transpose since it's a sequence
+
+    full_network = FullNetWork(lstm_network, fully_connected_network)
+
+Summary
+-------
+
+So we have created an LSTM network (``lstm_network``), a fully connected network (``fully_connected_network``), those two components are used in the full network. This full network used padded, packed sequences (defined in the forward pass), so we created the ``PadCollate`` class to process the need work. The DataLoader will conduct that process. Finally, when we load the data, this will be done using the vectorizer, so the address will be represented using word embeddings. Also, the address components will be converted into categorical value (from 0 to 7).
```
Comment (freud14) [inline docs/source/examples/tips_and_tricks.rst:307]: Change mention of PadCollate class here.

File: docs/source/examples/tips_and_tricks.rst [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+                        callbacks=callbacks,
+                        coloring={"text_color": "MAGENTA", "ratio_color":"BLUE"})
+
+
+Epoch metrics
+-------------
+
+It's also possible to used epoch metrics such as F1-score. You could also define your own epoch metric using the ``EpochMetric`` interface.
+
+Furthermore, you could also use the ``SKLearnMetrics`` wrapper to wrap a Scikit-learn metric as an epoch metric.
```
Comment (freud14) [inline docs/source/examples/tips_and_tricks.rst:441]:
```suggestion
Furthermore, you could also use the ``SKLearnMetrics`` wrapper to wrap a scikit-learn metric as an epoch metric.
```

File: docs/source/examples/train_with_policy_module.rst [hunk 9]
```diff
@@ -0,0 +1,145 @@
+.. role:: hidden
+    :class: hidden-section
+
+Train CIFAR with the ``policy`` module
+**************************************
+.. note:: See the notebook `here <https://github.com/GRAAL-Research/poutyne/blob/master/examples/policy_cifar_example.ipynb>`_
+
+Let's import all the needed packages.
+.. code-block:: python
```
Comment (freud14) [inline docs/source/examples/train_with_policy_module.rst:9]:
```suggestion
Let's import all the needed packages.

.. code-block:: python
```

File: docs/source/examples/train_with_policy_module.rst [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    import torch.optim as optim
+    from poutyne.framework import Model
+    from poutyne.framework import OptimizerPolicy, one_cycle_phases
+
+
+Training constants
+==================
+
+But first, let's set the training constants, the Cuda device used for training if one is present, we set the batch size (i.e. the number of elements to see before updating the model) and the epoch number (i.e. the number of times we see the full dataset).
+.. code-block:: python
```
Comment (freud14) [inline docs/source/examples/train_with_policy_module.rst:26]:
```suggestion
But first, let's set the training constants, the Cuda device used for training if one is present, we set the batch size (i.e. the number of elements to see before updating the model) and the epoch number (i.e. the number of times we see the full dataset).

.. code-block:: python
```

File: docs/source/examples/tips_and_tricks.rst [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+
+    full_network = FullNetWork(lstm_network, fully_connected_network)
+
+Summary
+-------
+
+So we have created an LSTM network (``lstm_network``), a fully connected network (``fully_connected_network``), those two components are used in the full network. This full network used padded, packed sequences (defined in the forward pass), so we created the ``PadCollate`` class to process the need work. The DataLoader will conduct that process. Finally, when we load the data, this will be done using the vectorizer, so the address will be represented using word embeddings. Also, the address components will be converted into categorical value (from 0 to 7).
+
+The Training Loop
+=================
```
Comment (freud14) [inline docs/source/examples/tips_and_tricks.rst:310]: I think that we should not put this section title and put the sections below it as first-level section title instead so that they appear in the left menu since it's what we want to show.

File: docs/source/examples/train_with_policy_module.rst [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    import torch.nn as nn
+    import torch.optim as optim
+    from poutyne.framework import Model
+    from poutyne.framework import OptimizerPolicy, one_cycle_phases
+
+
+Training constants
+==================
+
+But first, let's set the training constants, the Cuda device used for training if one is present, we set the batch size (i.e. the number of elements to see before updating the model) and the epoch number (i.e. the number of times we see the full dataset).
```
Comment (freud14) [inline docs/source/examples/train_with_policy_module.rst:25]: Modify as mentioned above.

File: docs/source/examples/tips_and_tricks.rst [hunk 13]
```diff
@@ -0,0 +1,492 @@
+.. role:: hidden
+    :class: hidden-section
+
+Poutyne's Tips and Tricks
```
Comment (freud14) [inline docs/source/examples/tips_and_tricks.rst:4]:
```suggestion
Tips and Tricks
```

File: docs/source/examples/transfert_learning.rst [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+.. code-block:: python
+
+    set_seeds(42)
+
+
+.. code-block:: python
+
+    def download_and_extract_dataset(path):
+        tgz_filename = "images.tgz"
+        urllib.request.urlretrieve("http://www.vision.caltech.edu/visipedia-data/CUB-200/images.tgz", tgz_filename)
```
Comment (freud14) [inline docs/source/examples/transfert_learning.rst:44]: Change URL for new URL

File: docs/source/examples/transfert_learning.rst [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+    valid_path = os.path.join(base_path, 'valid')
+    test_path = os.path.join(base_path, 'test')
+
+.. code-block:: python
+
+    download_and_extract_dataset(base_path)
+    split_train_valid_test(dataset_path, train_path, valid_path, test_path)
+
+
+Now, let's set our training constants. We first have the Cuda device used for training if one is present. Secondly, we set the number of classes (i.e. one for each number). Finally, we set the batch size (i.e. the number of elements to see before updating the model), the learning rate for the optimizer, and the epoch number (i.e. the number of times we see the full dataset).
```
Comment (freud14) [inline docs/source/examples/transfert_learning.rst:100]: Modify as mentioned above.

File: docs/source/examples/transfert_learning.rst (hunk 14 above)
Comment (davebulaval) [reply docs/source/examples/transfert_learning.rst:44]: Will be done in next PR.

File: docs/source/examples/tips_and_tricks.rst (hunk 5 above)
Comment (davebulaval) [reply docs/source/examples/tips_and_tricks.rst:217]: Yeah, I've misexplained. I will rewrite.

File: docs/source/examples/tips_and_tricks.rst (hunk 7 above)
Comment (davebulaval) [reply docs/source/examples/tips_and_tricks.rst:307]: Done

File: docs/source/examples/tips_and_tricks.rst (hunk 5 above)
Comment (davebulaval) [reply docs/source/examples/tips_and_tricks.rst]: 4b2c42d

**Comment by davebulaval** [comment]
Ajouter figure policy module
